### PR TITLE
apps sc: elasticsearch slm script deletes indices in bulk

### PR DESCRIPTION
**Pull request by viktor-f**
**What this PR does / why we need it**: We have had some problems with slm taking a long time to delete indices, hopefully it helps if we try to delete indices in bulk instead.

**Which issue this PR fixes** *(use the format `fixes #<issue number>(, fixes #<issue_number>, ...)` to automatically close the issue when PR gets merged)*: no issue

**Public facing documentation PR** *(if applicable)*

**Special notes for reviewer**:

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [x] is completely transparent, will not impact the workload in any way.
    - [ ] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.

**Pipeline config** *(if applicable)*
If you change some config options (e.g. add/rename variable or change the default value) you may need to update the config used by the pipeline in `pipeline/config`.

### CODE REVIEW

File: helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash [hunk 1]
```diff
@@ -89,29 +92,36 @@ function remove_old_snapshots {
...
 
     while [ $((snapshot_count - idx )) -gt "${MAX_SNAPSHOTS}" ]; do
         local snapshot_name
         snapshot_name=$(echo "${snapshots}" | jq -r ".[${idx}].snapshot")
         echo "too many snapshots: $((snapshot_count - idx )), max ${MAX_SNAPSHOTS}"
-        remove_snapshot "${snapshot_name}"
+        snapshots_to_delete="${snapshots_to_delete}${snapshot_name},"
         idx=$((idx + 1))
     done
+    if [ ! -z "${snapshots_to_delete}" ]; then
```
Comment (OlleLarsson) [inline helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash:122]:
```suggestion
    if [ -n "${snapshots_to_delete}" ]; then
```

File: helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash [hunk 2]
```diff
@@ -49,8 +49,9 @@ function get_snapshot_age {
 }
 
 function remove_snapshot {
```
Comment (OlleLarsson) [inline helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash:51]: Maybe we should rename this function to
```suggestion
function remove_snapshots {
```
?

File: helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash [hunk 3]
```diff
@@ -49,8 +49,9 @@ function get_snapshot_age {
...
-    local snapshot_name=$1
-    local url="${ELASTICSEARCH_URL}/_snapshot/${SNAPSHOT_REPOSITORY}/${snapshot_name}"
+    local snapshots_to_delete=$1
+    local url="${ELASTICSEARCH_URL}/_snapshot/${SNAPSHOT_REPOSITORY}/${snapshots_to_delete}"
+    echo "Snapshots to delete: ${snapshots_to_delete}"
```
Comment (OlleLarsson) [inline helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash:54]:
```suggestion
    echo "Deleting snapshots: ${snapshots_to_delete}"
```

**Review by OlleLarsson — COMMENTED** [review]
Looks good, i'll take a closer look in a while.

File: helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash [hunk 4]
```diff
@@ -75,12 +76,14 @@ function remove_old_snapshots {
...
     local snapshots
     local snapshot_count
+    local snapshots_to_delete=""
 
     snapshots=$(get_snapshots)
     snapshot_count=$(echo "${snapshots}" | jq length)
+    echo "Number of snapshots $snapshot_count"
 
-    check_snapshot_count "${snapshot_count}" || exit 0
-    check_old_snapshots  "${snapshots}"      || exit 0
```
Comment (lentzi90) [inline helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash:83]: Great! This was the bug I talked about also. :slightly_smiling_face:

File: helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash [hunk 5]
```diff
@@ -75,12 +76,14 @@ function remove_old_snapshots {
     local idx=0
     local snapshots
     local snapshot_count
+    local snapshots_to_delete=""
```
Comment (lentzi90) [inline helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash:79]: Could you add a log line here also to make it more obvious what's going on?
Something like
```bash
echo "Checking for old snapshots."
```

File: helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash [hunk 6]
```diff
@@ -89,29 +92,36 @@ function remove_old_snapshots {
...
+    if [ ! -z "${snapshots_to_delete}" ]; then
+        remove_snapshot "${snapshots_to_delete}"
+    fi
 }
 
 function remove_excess_snapshots {
     local idx=0
     local snapshots
     local snapshot_count
+    local snapshots_to_delete=""
```
Comment (lentzi90) [inline helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash:109]:
```suggestion
echo "Checking number of snapshots."
```

File: helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash [hunk 7]
```diff
@@ -89,29 +92,36 @@ function remove_old_snapshots {
...
-        remove_snapshot "${snapshot_name}"
+        snapshots_to_delete="${snapshots_to_delete}${snapshot_name},"
         idx=$((idx + 1))
     done
+    if [ ! -z "${snapshots_to_delete}" ]; then
+        remove_snapshot "${snapshots_to_delete}"
+    fi
 }
 
 remove_old_snapshots
```
Comment (lentzi90) [inline helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash:127]: Let's add a bit more logging here also so it is more obvious what it is doing:

```bash
echo "SLM retention procedure started"
echo "Removing snapshots older than ${MAX_AGE_SECONDS} seconds but keeping at least ${MIN_SNAPSHOTS} snapshots and at most ${MAX_SNAPSHOTS} snapshots."
remove_old_snapshots
remove_excess_snapshots
echo "SLM retention procedure done."
```

File: helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash [hunk 8]
```diff
@@ -75,12 +76,14 @@ function remove_old_snapshots {
     local idx=0
     local snapshots
     local snapshot_count
+    local snapshots_to_delete=""
 
     snapshots=$(get_snapshots)
     snapshot_count=$(echo "${snapshots}" | jq length)
+    echo "Number of snapshots $snapshot_count"
```
Comment (lentzi90) [inline helmfile/charts/elasticsearch-slm/scripts/slm-retention.bash:83]: nit: I would move this to `remove_excess_snapshots`. It is nice for debugging, but doesn't make sense when checking for age.
```suggestion
```

**Review by lentzi90 — COMMENTED** [review]
Great job!
I added some suggestions on how to make the logging a bit better.

**Review by lentzi90 — APPROVED** [review]
LGTM!
You could add another entry in the changelog for "fixed: SLM now deletes excess snapshots also when none of them are older than the maximum age" if you want.

**Review by OlleLarsson — APPROVED** [review]
lgtm

**Pull request merged**